gctree/dag_filtering.py

```diff
@@ -0,0 +1,507 @@
+r"""Functions for choosing optimal trees according to flexible criteria."""
```

Comment:
Do we want this all to be a public module exposed in the API and docs for end users, or is much of this better kept under the hood?

Reply:
I suppose the `filter_dag` function covers most of what someone might want to do. Having access to the filtering functions would make it easy to define custom criteria (e.g. some other branching process model likelihood, etc) and trim for that and an existing criterion jointly, or to trim according to the existing criteria (lexicographically) in a custom order. For example, you could do

```python
import historydag as hdag
import gctree.dag_filtering as df

with open('dagfile.p', 'rb') as fh:
    dag = hdag.deserialize(fh.read())

iso = df.isotype_dagfuncs(...)
mut = df.mutability_dagfuncs(...)
coolnewthing = hdag.utils.AddFuncDict(
    {
        # some functions...
    },
    names="coolnewthing"
)

dag.trim_optimal_weight(**(coolnewthing + iso + mut), optimal_func=min)
bp.CollapsedTree(dag.sampe().to_ete(...)).render(...)
```

This seems pretty slick to me, but that's not really justification for exposing it in the API, I guess. Maybe it's more in the scope of hDAG, and these filtering functions should be part of the hDAG api?

Do you think the current functionality of the `filter_dag` function is flexible enough?

Reply:
I agree that's slick, so I'll leave it to you to make the call on what should be public. I agree that more generic/abstract functionality could live in hdag, where you might want to reuse it.

Reply:
Part of the reason I ended up moving all this to gctree was that having it in hdag required I import gctree to avoid duplicating (or moving) a bunch of fundamental gctree code. Then we're into circular dependency territory, so I think I'd better leave it here.

I think I'll make it public, and work on some better documentation. Now that I have historydag docs up, that should be easier.

Reply:
Sounds good. Can you point me at historydag docs? I could't find a link in the readme.

Reply:
https://matsengrp.github.io/historydag/
Needs organization. Right now it's just built from docstrings with no context pages.

Reply:
Nice, thanks! You could link it in your README if you want it findable